### PR TITLE
[SimplifyCFG][JumpThreading] Don't duplicate inline-asm instructions

### PR DESCRIPTION
**Pull request by mcinally**
Conservatively, don't duplicate an inline-asm instruction during Threading. In this particular case, a label is defined in the inline-asm instruction. Duplicating that inline-asm instruction will result in 2 identical labels being defined, which causes an assembler error.

### CODE REVIEW

**Comment by yxsamliu** [comment]
A concern is that this may cause perf degradation for inline asm that allows duplication.

Does `call void asm sideeffect` prevent duplication?

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp [hunk 1]
```diff
@@ -491,8 +491,10 @@ static unsigned getJumpThreadDuplicationCost(const TargetTransformInfo *TTI,
 
     // Blocks with NoDuplicate are modelled as having infinite cost, so they
     // are never duplicated.
+    // Conservatively disallow inline-asm instructions. Duplicating inline-asm
+    // instructions can potentially create duplicate labels.
     if (const CallInst *CI = dyn_cast<CallInst>(I))
-      if (CI->cannotDuplicate() || CI->isConvergent())
+      if (CI->cannotDuplicate() || CI->isConvergent() || CI->isInlineAsm())
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/JumpThreading.cpp:497]: *If* we want to say that inline asm cannot be duplicated, then that should be part of `cannotDuplicate()`, not a separate query. There are a number of other places checking cannotDuplicate().

**Review by nickdesaulniers — CHANGES_REQUESTED** [review]
We've seen this in the Linux kernel before, pretty sure `\@` should be used on the label to make it unique per instantiation.  Let me see if I can find the kernel commit I'm thinking of.

(Cant find the kernel commit I'm thinking of).  Sorry `%=` is the sigil. This problem is why it exists.

> Under certain circumstances, GCC may duplicate (or remove duplicates of) your assembly code when optimizing. This can lead to unexpected duplicate symbol errors during compilation if your asm code defines symbols or labels. Using ‘%=’ (see [AssemblerTemplate](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#AssemblerTemplate)) may help resolve this problem.

`\@` is for asm .macro expansion, but same idea.
https://ftp.gnu.org/old-gnu/Manuals/gas-2.9.1/html_chapter/as_7.html#SEC109

**Comment by mcinally** [comment]
> We've seen this in the Linux kernel before, pretty sure `\@` should be used on the label to make it unique per instantiation. Let me see if I can find the kernel commit I'm thinking of.
> 
> (Cant find the kernel commit I'm thinking of). Sorry `%=` is the sigil. This problem is why it exists.
> 
> > Under certain circumstances, GCC may duplicate (or remove duplicates of) your assembly code when optimizing. This can lead to unexpected duplicate symbol errors during compilation if your asm code defines symbols or labels. Using ‘%=’ (see [AssemblerTemplate](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#AssemblerTemplate)) may help resolve this problem.
> 
> `\@` is for asm .macro expansion, but same idea. https://ftp.gnu.org/old-gnu/Manuals/gas-2.9.1/html_chapter/as_7.html#SEC109

Ah, that's compelling. Thanks, Nick!

Closing this issue...

**Pull request closed without merging**